packages/million/block.ts

```diff
@@ -0,0 +1,262 @@
+import {
+  childNodes$,
+  cloneNode$,
+  createEventListener,
+  insertBefore$,
+  insertText,
+  innerHTML$,
+  remove$ as removeElement$,
+  setAttribute,
+  setText,
+  mapGet$,
+  mapHas$,
+  mapSet$,
+  setStyleAttribute,
+  setSvgAttribute,
+  firstChild$,
+  nextSibling$,
+} from './dom';
+import { renderToTemplate } from './template';
+import { AbstractBlock } from './types';
+import type { Edit, EditChild, Props, VElement, Hole, VNode } from './types';
+
+export const createBlock = (
+  fn: (props?: Props) => VElement,
+  unwrap?: (vnode: any) => VNode,
+) => {
+  const vnode = fn(
+    new Proxy(
+      {},
+      {
+        // A universal getter will return a Hole instance if props[any] is accessed
+        // Allows code to identify holes in virtual nodes ("digs" them out)
+        get(_, key: string): Hole {
+          return { __key: key };
+        },
+      },
+    ),
+  );
+  const edits: Edit[] = [];
+
+  // Turns vnode into a string of HTML and creates an array of "edits"
+  // Edits are instructions for how to update the DOM given some props
+  const root = stringToDOM(
+    renderToTemplate(unwrap ? unwrap(vnode) : vnode, edits),
+  );
+
+  // Handles case for positioning text nodes. When text nodes are
+  // put into a template, they can be merged. For example,
+  // ["hello", "world"] becomes "helloworld" in the DOM.
+  // Inserts text nodes into the DOM at the correct position.
+  for (let i = 0, j = edits.length; i < j; ++i) {
+    const current = edits[i]!;
+    const initsLength = current.inits.length;
+    if (!initsLength) continue;
+    const el = getCurrentElement(current, root);
+    for (let k = 0; k < initsLength; ++k) {
+      const init = current.inits[k]!;
+      insertText(el, init.value, init.index);
+    }
+  }
+
+  return (
+    props?: Props | null,
+    key?: string,
+    shouldUpdate?: (oldProps: Props, newProps: Props) => boolean,
+  ) => {
+    return new Block(root, edits, props, key ?? props?.key, shouldUpdate);
+  };
+};
+
+export class Block extends AbstractBlock {
+  root: HTMLElement;
+  edits: Edit[];
+  // Cache for getCurrentElement()
+  cache = new Map<number, HTMLElement>();
+
+  constructor(
+    root: HTMLElement,
+    edits: Edit[],
+    props?: Props | null,
+    key?: string,
+    shouldUpdate?: (oldProps: Props, newProps: Props) => boolean,
+  ) {
+    super();
+    this.root = root;
+    this.props = props;
+    this.edits = edits;
+    this.key = key;
+    if (shouldUpdate) this.shouldUpdate = shouldUpdate;
+  }
+  mount(
+    parent?: HTMLElement,
+    refNode: Node | null = null,
+    inPlace = false,
+  ): HTMLElement {
+    if (this.el) return this.el;
+    // cloneNode(true) uses less memory than recursively creating new nodes
+    const root = cloneNode$.call(this.root, true) as HTMLElement;
+
+    for (let i = 0, j = this.edits.length; i < j; ++i) {
+      const current = this.edits[i]!;
+      const el = getCurrentElement(current, root, this.cache, i);
```

Comment:
Continues from https://github.com/aidenybai/million/pull/292#discussion_r1118877216

> `edits` act like a blueprint for an instance of a Block to reference to update based on props.
> 
> [(click on reference)](https://github.com/aidenybai/million/blob/3ebd41e599be86d572bba87f54fa2e88b19e0ddc/packages/million/block.ts#L45) An empty `edits` array is passed into the `renderToTemplate` code. Inside that function, the edits array is mutated ([ref](https://github.com/aidenybai/million/blob/3ebd41e599be86d572bba87f54fa2e88b19e0ddc/packages/million/template.ts#L170)). Later on, the `edits`array will be passed into a new Block instance ([code](https://github.com/aidenybai/million/blob/3ebd41e599be86d572bba87f54fa2e88b19e0ddc/packages/million/block.ts#L68))
> 
> The `edits` will never be changed on mount/patch. However, during mount/patch it is iterated through to apply edits to the Block's element (`this.el`)

Since the `current` here could be considered immutable, we can use `current` as the cache key (`cache.has(current)`) instead of the numeric index. And then we can use `WeakMap` (instead of `Map`) for caching, then the cached HTMLElement could be released if `current` ever get GC-ed.

Reply:
Seems like there is not a big difference in memory usage after bench, but change made anyway